tests/e2e/models/mnist_train_test/mnist_train_test.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from iree.runtime import load_vm_flatbuffer_file
+from iree.compiler.tools import compile_file, InputType
+import os
+import sys
+import tempfile
+import numpy as np
+import unittest
+import argparse
+from typing import Any, Callable, List
+
+args = None
+
+
+def build_module():
+  with tempfile.TemporaryDirectory() as tmp_dir:
+    vmfb_file = os.path.join(tmp_dir, "mnist_train.vmfb")
+    compile_file(input_file=os.path.join(os.path.dirname(__file__),
+                                         "mnist_train.mlir"),
+                 output_file=vmfb_file,
+                 target_backends=[args.target_backend],
+                 input_type=InputType.MHLO)
+    return load_vm_flatbuffer_file(vmfb_file, driver=args.driver)
+
+
+def load_data():
+  data_dir = os.path.dirname(__file__)
+  batch = list(np.load(os.path.join(data_dir, "batch.npz")).values())
+  expected_optimizer_state_after_init = list(
+      np.load(os.path.join(data_dir,
+                           "expected_optimizer_state_after_init.npz")).values())
+  expected_optimizer_state_after_train_step = list(
+      np.load(
+          os.path.join(
+              data_dir,
+              "expected_optimizer_state_after_train_step.npz")).values())
+  expected_prediction_after_train_step = list(
+      np.load(os.path.join(
+          data_dir, "expected_prediction_after_train_step.npz")).values())[0]
+  return batch, expected_optimizer_state_after_init, expected_optimizer_state_after_train_step, expected_prediction_after_train_step
+
+
+def parse_args(args: List[str] = sys.argv[1:]):
+  parser = argparse.ArgumentParser()
+  parser.add_argument("--target_backend", type=str, default="llvm-cpu")
+  parser.add_argument("--driver", type=str, default="local-task")
+  return parser.parse_known_args(args=args)
+
+
+DEFAULT_DECIMAL = 5
+DEFAULT_EPSILON = 10**-7
+DEFAULT_NULP = 10**8
+
+
+def assert_array_almost_equal(a,
+                              b,
+                              decimal=DEFAULT_DECIMAL,
+                              epsilon=DEFAULT_EPSILON,
+                              nulp=DEFAULT_NULP):
+  np_a = np.asarray(a)
+  np_b = np.asarray(b)
+  # Test for absolute error.
+  np.testing.assert_array_almost_equal(np_a, np_b, decimal=decimal)
+  # Test for relative error while ignoring false errors from
+  # catastrophic cancellation.
```

Comment:
Is there a nice link with an explanation for this? The docs on `assert_array_almost_equal_nulp` suggest that it's a replacement for `assert_array_almost_equal` and that it would already be doing this comparison without the zeros and looking up "catastrophic cancellation" suggests that it's the subtraction itself here that is introducing that possibility, whereas preserving the magnitudes would allow reasoning about the error in context. In fact, this looks like almost the opposite of relative errors, since that's based on the ratio of the difference to the overall magnitude and this is throwing out the magnitude

Reply:
Yes. This is wrong. I reworked array comparison to use `numpy.allclose`.